omegaconf/base.py

```diff
@@ -325,117 +311,194 @@ def _select_impl(
-        if inter_type == "str":
-            root_node, inter_key = self._resolve_key_and_root(inter_key)
-            parent, last_key, value = root_node._select_impl(
-                inter_key,
-                throw_on_missing=throw_on_missing,
-                throw_on_resolution_failure=throw_on_resolution_failure,
-            )
-
-            # if parent is None or (value is None and last_key not in parent):  # type: ignore
-            if parent is None or value is None:
-                if throw_on_resolution_failure:
-                    raise ConfigKeyError(
-                        f"{inter_type} interpolation key '{inter_key}' not found"
-                    )
-                else:
-                    return None
-            assert isinstance(value, Node)
-            return value
+        resolver = OmegaConf.get_resolver(inter_type)
+        if resolver is not None:
+            root_node = self._get_root()
+            try:
+                value = resolver(root_node, inter_args, inter_args_str)
+                return ValueNode(
+                    value=value,
+                    parent=self,
+                    metadata=Metadata(
+                        ref_type=None, object_type=None, key=key, optional=True
+                    ),
+                )
+            except Exception as e:
+                self._format_and_raise(key=inter_args, value=None, cause=e)
+                assert False
         else:
-            resolver = OmegaConf.get_resolver(inter_type)
-            if resolver is not None:
-                root_node = self._get_root()
-                try:
-                    value = resolver(root_node, inter_key)
-                    return ValueNode(
-                        value=value,
-                        parent=self,
-                        metadata=Metadata(
-                            ref_type=None, object_type=None, key=key, optional=True
-                        ),
-                    )
-                except Exception as e:
-                    self._format_and_raise(key=inter_key, value=None, cause=e)
-                    assert False
+            if throw_on_resolution_failure:
+                raise UnsupportedInterpolationType(
+                    f"Unsupported interpolation type {inter_type}"
+                )
             else:
-                if throw_on_resolution_failure:
-                    raise UnsupportedInterpolationType(
-                        f"Unsupported interpolation type {inter_type}"
-                    )
-                else:
-                    return None
+                return None
 
     def _resolve_interpolation(
         self,
+        parent: Optional["Container"],
         key: Any,
         value: "Node",
         throw_on_missing: bool,
         throw_on_resolution_failure: bool,
     ) -> Any:
+        value_kind = get_value_kind(value)
+        if value_kind != ValueKind.INTERPOLATION:
+            return value
```

Comment:
should this function ever be called without a check that the value is an interpolation?

Reply:
Then we'd have to add that check everywhere where it's called. What about renaming it into `_maybe_resolve_interpolation()`?

Reply:
> Then we'd have to add that check everywhere it's called. What about renaming it into `_maybe_resolve_interpolation()`?

I did this rename in 2ad3dca

This made me realize that there was one place where this function was called after having already checked that it is an interpolation, so we were doing two checks for no good reason. This has been addressed in cc834cb